File: euler28.py
# Начиная с числа 1 и двигаясь дальше вправо по часовой стрелке,
# считаем суммы диагоналей. Какова сумма чисел в диагоналях спирали
# 1001 на 1001, образованной таким же способом?

from project_euler_defs import *

n = 1001

# 1 вариант
start_time = datetime.now()
num = 1
print(sum(num := num + lyr for lyr in range(2, n, 2) for _ in range(4)) + 1)
print(datetime.now() - start_time)

# 2 вариант
start_time = datetime.now()
print(sum(map(lambda i: 4 * i ** 2 - 6 * (i - 1), range(3, n + 1, 2))) + 1)
print(datetime.now() - start_time)

# 3 вариант
start_time = datetime.now()
print((4 * n ** 3 + 3 * n ** 2 + 8 * n - 9) // 6)
print(datetime.now() - start_time)

File: go/euler28.go
// euler28 находит сумму чисел в диагоналях спирали 1001×1001.
//
// Решение задачи Эйлера №28.
package main

import (
	"fmt"
	"time"
)

func euler28() {
	n := 1001

	startTime := time.Now()
	num := 1
	result := 0
	lyr := 2
	for lyr <= n {
		for i := 0; i < 4; i++ {
			num += lyr
			result += num
		}
		lyr += 2
	}
	result++
	fmt.Println(result)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	sum := 1
	for i := 3; i <= n; i += 2 {
		sum += 4*i*i - 6*(i-1)
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	res := (4*n*n*n + 3*n*n + 8*n - 9) / 6
	fmt.Println(res)
	fmt.Println(time.Since(startTime))
}

func main() {
	euler28()
}
